File: source/rust/src/BoundingBox/rect.rs
#[repr(C)]
pub struct PointI32 {
    pub x: i32,
    pub y: i32
}

#[repr(C)]
pub struct Rect{
    pub x : u16,
    pub y: u16,
    pub width: u16, 
    pub height: u16,
    pub center: PointI32
}

impl Rect {

    pub fn intersectingRect(&self, other: &Rect) -> Rect {

        let x: u16 = self.x.max(other.x);
        let y: u16 = self.y.max(other.y);
        let width: u16 = (self.x + self.width).min(other.x + other.width) - x;
        let height: u16 = (self.y + self.height).min(other.y + other.height) - y;  
        let center: PointI32 = PointI32 { x: (x + (width / 2)) as i32, y: (y + (height / 2)) as i32 }; 

        Rect { x, y, width, height, center } 

    }

    pub fn area(&self) -> u32{
        self.height as u32 * self.width as u32
    }

    pub fn intersectionOfUnionPercentage(&self, other: &Rect ) -> f32{
        let intersectingRectArea:u32 = self.intersectingRect(other).area();
        let unionArea: u32 = self.area() + other.area() - intersectingRectArea;
        println!("Intersecting Area: {}", intersectingRectArea);
        println!("Union Area: {}", unionArea);
        intersectingRectArea as f32 / unionArea as f32

    }

}

File: source/c++/src/rect.cpp
#include "rect.hpp"

using std::cout;

uint Rect::squareDistanceTo(const Rect& r) const {
    return squareDistance(this->center, r.center);
}

bool Rect::isCloseTo(Rect& r, double distance){
	return this->squareDistanceTo(r) <  distance;
}

void Rect::draw(cv::Mat& frame, cv::Scalar color){
    cv::rectangle(frame, *this, color, 2);
}

vector<Rect> pointsToRects(int32_t *points, uint16_t size){
	
	std::vector<Rect> rects;
    rects.reserve(size);
	for (uint16_t i = 0; i < size; i++){
        int32_t* currentPoint = points + i*core::_elementsInPoint;
        Point A = Point(currentPoint[0], currentPoint[1]);
        Point B = Point(currentPoint[2], currentPoint[3]);
        rects.emplace_back(A, B);
	}
    
	return rects;  

}

float32 Rect::iouPercentage(const Rect& other) const {
    int Xoverlap = MIN((this->x + this->width), (other.x + other.width)) - MAX(this->x, other.x);
    int Yoverlap = MIN((this->y + this->height), (other.y + other.height)) - MAX(this->y, other.y);
    if (Xoverlap <= 0.0 || Yoverlap <= 0.0) {
        return 0.0;
    }
    float32 intersectionArea = Xoverlap * Yoverlap;

    return intersectionArea / (float32) this->area();
}

std::ostream& operator<<(std::ostream& os, Rect& rect){
    // Printing all the elements
    os << "[" << rect.x << " " << rect.y << " " << rect.width << " " << rect.height << "]";
    os << "Area: " << rect.area() << "\n"; 
    return os;
}

